models/classes/runner/time/storageFormat/QtiTimeStorageJsonFormat.php

```diff
@@ -60,11 +61,16 @@ public function decode($data)
 
         if (is_array($decodedData)) {
             foreach ($decodedData as $key => &$value) {
```

Comment:
it is not clear for me:
1. if $value is an object why do you need it as a link?
2. why do we need to foreach all array, if we know what do we need, we can use something like:
if (is_array($decodedData) {
  $decodedData = $this->decodeTimeLIne($decodedData);
  $decodedData = $this->decodeTimerAdjuctmentMap($dcodedData)
}


...
function decodeTimeLIne (array $decodedData): array
{
if (array_key_exists(self::STORAGE_KEY_TIME_LINE, $decodedData) && !$decodedData[self::STORAGE_KEY_TIME_LINE] instanceof QtiTimeLine) {// do your encoding}
return $decodedData;
}

Reply:
Extracted a method to a trait following your suggestion